osu.Game/Rulesets/Difficulty/Utils/DifficultyHitObjectIterator.cs

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using osu.Game.Rulesets.Difficulty.Preprocessing;
+
+namespace osu.Game.Rulesets.Difficulty.Utils
+{
+    /// <summary>
+    /// Wraps a list of <see cref="DifficultyHitObject"/>s.
+    /// Provides access to the current object in time in difficulty calculation, as well as it's preceding and succeeding objects using relative indices.
+    /// </summary>
+    public class DifficultyHitObjectIterator : IEnumerator
+    {
+        private readonly List<DifficultyHitObject> difficultyHitObjects;
```

Comment:
This feels weird to me - you're storing the list you're working on inside the enumerator when it should be the other way around? I'm also not sure if the enumerator is even required here - why not make a `class DifficultyHitObjectsCollection : IList<DifficultyHitObject>`?

Reply:
Well, that's how `IEnumerator`s usually work, they receive a list and iterate on it using `MoveNext`. This class provides an additional feature of peeking at an object a number of steps back/forward, without moving the position of the iterator.